File: src/github.zig
const std = @import("std");

pub const PullRequestStatus = enum {
    open,
    closed,
    merged,
};

pub const PullRequest = struct {
    number: i64,
    title: []const u8,
    status: PullRequestStatus,
    base_branch: []const u8,
    merge_commit_sha: ?[]const u8,
    merged_at: ?[]const u8,

    pub fn deinit(self: *PullRequest, allocator: std.mem.Allocator) void {
        allocator.free(self.title);
        allocator.free(self.base_branch);
        if (self.merge_commit_sha) |sha| allocator.free(sha);
        if (self.merged_at) |date| allocator.free(date);
    }
};

const json_options: std.json.ParseOptions = .{ .ignore_unknown_fields = true };

const SearchResponse = struct {
    items: []const struct { number: i64 },
};

const CompareResponse = struct {
    status: []const u8,
};

const PullResponse = struct {
    number: i64,
    title: []const u8,
    state: []const u8,
    merged: bool,
    merge_commit_sha: ?[]const u8,
    merged_at: ?[]const u8,
    base: struct { ref: []const u8 },
};

pub const GitHubClient = struct {
    allocator: std.mem.Allocator,
    io: std.Io,
    environ_map: *const std.process.Environ.Map,

    pub const ValidationError = error{
        GhNotInstalled,
        GhNotAuthenticated,
    };

    pub fn init(
        allocator: std.mem.Allocator,
        io: std.Io,
        environ_map: *std.process.Environ.Map,
    ) !GitHubClient {
        try sanitizeEnv(environ_map);
        return .{
            .allocator = allocator,
            .io = io,
            .environ_map = environ_map,
        };
    }

    pub fn validate(self: *GitHubClient) ValidationError!void {
        if (!self.commandSucceeds(&.{ "gh", "--version" })) return error.GhNotInstalled;
        if (!self.commandSucceeds(&.{ "gh", "auth", "status" })) return error.GhNotAuthenticated;
    }

    pub fn searchPRsByChangedFiles(
        self: *GitHubClient,
        package_name: []const u8,
        days: u32,
    ) ![]PullRequest {
        const now = std.Io.Clock.real.now(self.io).toSeconds();
        const days_ago = now - (@as(i64, days) * std.time.s_per_day);

        var date_buffer: [20]u8 = undefined;
        const date = try formatIsoDate(&date_buffer, days_ago);

        var endpoint: std.Io.Writer.Allocating = .init(self.allocator);
        defer endpoint.deinit();

        try endpoint.writer.writeAll("/search/issues?q=");
        try writeUrlEncoded(&endpoint.writer, "repo:NixOS/nixpkgs ");
        try writeUrlEncoded(&endpoint.writer, package_name);
        try writeUrlEncoded(&endpoint.writer, " type:pr created:>");
        try writeUrlEncoded(&endpoint.writer, date);
        try endpoint.writer.writeAll("&per_page=100&sort=created&order=desc");

        const json_response = try self.makeRequest(endpoint.written());
        defer self.allocator.free(json_response);

        return try self.parseSearchResults(json_response);
    }

    pub fn checkCommitInBranch(
        self: *GitHubClient,
        commit_sha: []const u8,
        branch: []const u8,
    ) !bool {
        var endpoint_buffer: [256]u8 = undefined;
        const endpoint = std.fmt.bufPrint(
            &endpoint_buffer,
            "/repos/NixOS/nixpkgs/compare/{s}...{s}",
            .{ commit_sha, branch },
        ) catch return false;

        const json_response = self.makeRequest(endpoint) catch return false;
        defer self.allocator.free(json_response);

        const parsed = try std.json.parseFromSlice(CompareResponse, self.allocator, json_response, json_options);
        defer parsed.deinit();

        return std.mem.eql(u8, parsed.value.status, "ahead") or
            std.mem.eql(u8, parsed.value.status, "identical");
    }

    pub fn branchesContainingCommit(
        self: *GitHubClient,
        candidates: []const []const u8,
        commit_sha: []const u8,
    ) ![][]const u8 {
        var result = try std.ArrayList([]const u8).initCapacity(self.allocator, candidates.len);
        errdefer {
            for (result.items) |item| self.allocator.free(item);
            result.deinit(self.allocator);
        }

        for (candidates) |branch| {
            if (self.checkCommitInBranch(commit_sha, branch) catch false) {
                try result.append(self.allocator, try self.allocator.dupe(u8, branch));
            }
        }

        return try result.toOwnedSlice(self.allocator);
    }

    fn commandSucceeds(self: *GitHubClient, argv: []const []const u8) bool {
        const result = std.process.run(self.allocator, self.io, .{
            .argv = argv,
            .environ_map = self.environ_map,
            .stdout_limit = .limited(4096),
            .stderr_limit = .limited(4096),
        }) catch return false;
        defer self.allocator.free(result.stdout);
        defer self.allocator.free(result.stderr);

        return childSucceeded(result.term);
    }

    fn makeRequest(self: *GitHubClient, endpoint: []const u8) ![]u8 {
        const result = std.process.run(self.allocator, self.io, .{
            .argv = &.{
                "gh",
                "api",
                "-H",
                "Accept: application/vnd.github+json",
                "-H",
                "X-GitHub-Api-Version: 2022-11-28",
                endpoint,
            },
            .environ_map = self.environ_map,
            .stdout_limit = .limited(10 * 1024 * 1024),
            .stderr_limit = .limited(64 * 1024),
        }) catch return error.GitHubRequestFailed;
        errdefer self.allocator.free(result.stdout);
        defer self.allocator.free(result.stderr);

        if (!childSucceeded(result.term)) return error.GitHubRequestFailed;
        return result.stdout;
    }

    fn parseSearchResults(self: *GitHubClient, json: []const u8) ![]PullRequest {
        const parsed = try std.json.parseFromSlice(SearchResponse, self.allocator, json, json_options);
        defer parsed.deinit();

        var result = try std.ArrayList(PullRequest).initCapacity(self.allocator, parsed.value.items.len);
        errdefer {
            for (result.items) |*pr| pr.deinit(self.allocator);
            result.deinit(self.allocator);
        }

        for (parsed.value.items) |item| {
            try result.append(self.allocator, try self.getPRDetails(item.number));
        }

        return try result.toOwnedSlice(self.allocator);
    }

    fn getPRDetails(self: *GitHubClient, pr_number: i64) !PullRequest {
        var endpoint_buffer: [64]u8 = undefined;
        const endpoint = try std.fmt.bufPrint(
            &endpoint_buffer,
            "/repos/NixOS/nixpkgs/pulls/{d}",
            .{pr_number},
        );

        const json_response = try self.makeRequest(endpoint);
        defer self.allocator.free(json_response);

        const parsed = try std.json.parseFromSlice(PullResponse, self.allocator, json_response, json_options);
        defer parsed.deinit();

        const pr = parsed.value;
        const title = try self.allocator.dupe(u8, pr.title);
        errdefer self.allocator.free(title);

        const base_branch = try self.allocator.dupe(u8, pr.base.ref);
        errdefer self.allocator.free(base_branch);

        const status: PullRequestStatus = if (pr.merged)
            .merged
        else if (std.mem.eql(u8, pr.state, "open"))
            .open
        else
            .closed;

        const merge_commit_sha = if (pr.merge_commit_sha) |sha| try self.allocator.dupe(u8, sha) else null;
        errdefer if (merge_commit_sha) |sha| self.allocator.free(sha);

        const merged_at = if (pr.merged_at) |date| try self.allocator.dupe(u8, date) else null;
        errdefer if (merged_at) |date| self.allocator.free(date);

        return .{
            .number = pr.number,
            .title = title,
            .status = status,
            .base_branch = base_branch,
            .merge_commit_sha = merge_commit_sha,
            .merged_at = merged_at,
        };
    }
};

fn childSucceeded(term: std.process.Child.Term) bool {
    return switch (term) {
        .exited => |code| code == 0,
        else => false,
    };
}

fn sanitizeEnv(env: *std.process.Environ.Map) !void {
    _ = env.swapRemove("GH_FORCE_TTY");
    _ = env.swapRemove("FORCE_COLOR");

    try env.put("NO_COLOR", "1");
    try env.put("CLICOLOR", "0");
    try env.put("CLICOLOR_FORCE", "0");
    try env.put("GH_NO_UPDATE_NOTIFIER", "1");
    try env.put("GH_PAGER", "cat");
}

fn formatIsoDate(buffer: []u8, timestamp: i64) ![]const u8 {
    const epoch_seconds = std.time.epoch.EpochSeconds{ .secs = @intCast(timestamp) };
    const day_seconds = epoch_seconds.getDaySeconds();
    const year_day = epoch_seconds.getEpochDay().calculateYearDay();
    const month_day = year_day.calculateMonthDay();

    return try std.fmt.bufPrint(
        buffer,
        "{d:0>4}-{d:0>2}-{d:0>2}T{d:0>2}:{d:0>2}:{d:0>2}Z",
        .{
            year_day.year,
            month_day.month.numeric(),
            month_day.day_index + 1,
            day_seconds.getHoursIntoDay(),
            day_seconds.getMinutesIntoHour(),
            day_seconds.getSecondsIntoMinute(),
        },
    );
}

fn writeUrlEncoded(writer: *std.Io.Writer, input: []const u8) !void {
    const hex = "0123456789ABCDEF";

    for (input) |c| {
        if (std.ascii.isAlphanumeric(c) or c == '-' or c == '_' or c == '.' or c == '~') {
            try writer.writeByte(c);
        } else if (c == ' ') {
            try writer.writeByte('+');
        } else {
            try writer.writeByte('%');
            try writer.writeByte(hex[c >> 4]);
            try writer.writeByte(hex[c & 0x0f]);
        }
    }
}

File: src/github.rs
use std::thread;

use anyhow::{
  Context,
  Result,
  bail,
};
use chrono::{
  Duration,
  SecondsFormat,
  Utc,
};
use serde::{
  Deserialize,
  de::DeserializeOwned,
};

mod auth;
mod transport;

pub use transport::GhCliStatus;
use transport::Transport;

const MAX_PARALLEL_REQUESTS: usize = 8;
const PULL_REQUEST_URL_BASE: &str = "https://github.com/NixOS/nixpkgs/pull";

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum PullRequestState {
  Open,
  Closed,
  Merged { merge_commit_sha: String },
}

impl PullRequestState {
  pub fn merge_commit_sha(&self) -> Option<&str> {
    match self {
      Self::Merged { merge_commit_sha } => Some(merge_commit_sha),
      Self::Open | Self::Closed => None,
    }
  }
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct PullRequest {
  pub number:      i64,
  pub title:       String,
  pub state:       PullRequestState,
  pub base_branch: String,
}

impl PullRequest {
  pub fn url(&self) -> String {
    format!("{PULL_REQUEST_URL_BASE}/{}", self.number)
  }
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct BranchReachability {
  pub branch: String,
  pub status: BranchReachabilityStatus,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct BranchReachabilityRequest {
  pub branch:     String,
  pub commit_sha: String,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum BranchReachabilityStatus {
  Contains,
  Missing,
  Unknown(String),
}

pub struct GitHubClient {
  transport: Transport,
}

pub enum GitHubAuth {
  GhCli,
  Token(String),
}

#[derive(Deserialize)]
struct SearchResponse {
  items: Vec<SearchItem>,
}

#[derive(Deserialize)]
struct SearchItem {
  number: i64,
}

#[derive(Deserialize)]
struct CompareResponse {
  status: String,
}

#[derive(Deserialize)]
struct PullResponse {
  number:           i64,
  title:            String,
  state:            String,
  merged:           bool,
  merge_commit_sha: Option<String>,
  base:             PullBase,
}

#[derive(Deserialize)]
struct PullBase {
  #[serde(rename = "ref")]
  ref_name: String,
}

pub fn gh_cli_status() -> GhCliStatus {
  transport::gh_cli_status()
}

pub fn load_token_after_unavailable_gh() -> Result<String> {
  auth::load_token(auth::TokenFallbackReason::GhUnavailable)
}

pub fn load_token_after_unauthenticated_gh() -> Result<String> {
  auth::load_token(auth::TokenFallbackReason::GhUnauthenticated)
}

impl GitHubClient {
  pub fn new(auth: GitHubAuth) -> Result<Self> {
    Ok(Self {
      transport: Transport::new(auth)?,
    })
  }

  pub fn search_prs_by_changed_files(
    &self,
    package_name: &str,
    days: u32,
  ) -> Result<Vec<PullRequest>> {
    let date = (Utc::now() - Duration::days(i64::from(days)))
      .to_rfc3339_opts(SecondsFormat::Secs, true);
    let query =
      format!("repo:NixOS/nixpkgs {package_name} type:pr created:>{date}");
    let endpoint = format!(
      "/search/issues?q={}&per_page=100&sort=created&order=desc",
      url_encode(&query)
    );

    let response: SearchResponse = self.get_json(&endpoint)?;
    let numbers = response
      .items
      .into_iter()
      .map(|item| item.number)
      .collect::<Vec<_>>();

    self.get_pr_details_parallel(&numbers)
  }

  pub fn probe_branch_reachability(
    &self,
    requests: &[BranchReachabilityRequest],
  ) -> Vec<BranchReachability> {
    bounded_parallel_map(requests, |request| {
      let status = match self
        .check_commit_in_branch(&request.commit_sha, &request.branch)
      {
        Ok(true) => BranchReachabilityStatus::Contains,
        Ok(false) => BranchReachabilityStatus::Missing,
        Err(err) => BranchReachabilityStatus::Unknown(err.to_string()),
      };

      BranchReachability {
        branch: request.branch.clone(),
        status,
      }
    })
  }

  fn get_pr_details_parallel(
    &self,
    numbers: &[i64],
  ) -> Result<Vec<PullRequest>> {
    bounded_parallel_map(numbers, |number| self.get_pr_details(*number))
      .into_iter()
      .collect()
  }

  fn check_commit_in_branch(
    &self,
    commit_sha: &str,
    branch: &str,
  ) -> Result<bool> {
    let endpoint =
      format!("/repos/NixOS/nixpkgs/compare/{commit_sha}...{branch}");
    let response: CompareResponse = self.get_json(&endpoint)?;

    Ok(response.status == "ahead" || response.status == "identical")
  }

  fn get_pr_details(&self, pr_number: i64) -> Result<PullRequest> {
    let endpoint = format!("/repos/NixOS/nixpkgs/pulls/{pr_number}");
    let pr: PullResponse = self.get_json(&endpoint)?;
    let state = pull_request_state(&pr)
      .with_context(|| format!("invalid GitHub PR payload for #{pr_number}"))?;

    Ok(PullRequest {
      number: pr.number,
      title: pr.title,
      state,
      base_branch: pr.base.ref_name,
    })
  }

  fn get_json<T>(&self, endpoint: &str) -> Result<T>
  where
    T: DeserializeOwned,
  {
    let body = self.transport.make_request(endpoint)?;
    serde_json::from_str(&body).with_context(|| {
      format!("failed to parse GitHub response from {endpoint}")
    })
  }
}

fn pull_request_state(pr: &PullResponse) -> Result<PullRequestState> {
  if pr.merged {
    let merge_commit_sha = pr
      .merge_commit_sha
      .clone()
      .context("merged PR has no merge_commit_sha")?;

    return Ok(PullRequestState::Merged { merge_commit_sha });
  }

  if pr.state == "open" {
    Ok(PullRequestState::Open)
  } else if pr.state == "closed" {
    Ok(PullRequestState::Closed)
  } else {
    bail!("unknown PR state {}", pr.state)
  }
}

fn url_encode(input: &str) -> String {
  const HEX: &[u8; 16] = b"0123456789ABCDEF";
  let mut encoded = String::with_capacity(input.len());

  for byte in input.bytes() {
    if byte.is_ascii_alphanumeric() || matches!(byte, b'-' | b'_' | b'.' | b'~')
    {
      encoded.push(char::from(byte));
    } else if byte == b' ' {
      encoded.push('+');
    } else {
      encoded.push('%');
      encoded.push(char::from(HEX[(byte >> 4) as usize]));
      encoded.push(char::from(HEX[(byte & 0x0F) as usize]));
    }
  }

  encoded
}

fn bounded_parallel_map<T, R, F>(items: &[T], f: F) -> Vec<R>
where
  T: Sync,
  R: Send,
  F: Fn(&T) -> R + Sync,
{
  let mut results = Vec::with_capacity(items.len());
  let f = &f;

  for chunk in items.chunks(MAX_PARALLEL_REQUESTS) {
    let mut chunk_results = thread::scope(|scope| {
      let mut handles = Vec::with_capacity(chunk.len());
      for item in chunk {
        handles.push(scope.spawn(move || f(item)));
      }

      handles
        .into_iter()
        .map(|handle| {
          handle.join().unwrap_or_else(|payload| {
            std::panic::resume_unwind(payload);
          })
        })
        .collect::<Vec<_>>()
    });

    results.append(&mut chunk_results);
  }

  results
}

#[cfg(test)]
mod tests {
  use super::{
    PullBase,
    PullRequestState,
    PullResponse,
    pull_request_state,
    url_encode,
  };

  #[test]
  fn encodes_github_search_query_like_gh_endpoint() {
    assert_eq!(
      "repo%3ANixOS%2Fnixpkgs+zig+type%3Apr+created%3A%3E2026-01-01T00%3A00%\
       3A00Z",
      url_encode(
        "repo:NixOS/nixpkgs zig type:pr created:>2026-01-01T00:00:00Z"
      )
    );
  }

  #[test]
  fn merged_pr_state_requires_a_merge_commit_sha() {
    let response = pull_response(true, "closed", None);

    assert!(pull_request_state(&response).is_err());
  }

  #[test]
  fn merged_pr_state_owns_the_merge_commit_sha() {
    let response = pull_response(true, "closed", Some("abc123"));

    assert_eq!(
      Some(PullRequestState::Merged {
        merge_commit_sha: "abc123".to_string(),
      }),
      pull_request_state(&response).ok()
    );
  }

  fn pull_response(
    merged: bool,
    state: &str,
    merge_commit_sha: Option<&str>,
  ) -> PullResponse {
    PullResponse {
      number: 1,
      title: "title".to_string(),
      state: state.to_string(),
      merged,
      merge_commit_sha: merge_commit_sha.map(str::to_string),
      base: PullBase {
        ref_name: "master".to_string(),
      },
    }
  }
}
